resolwe/storage/manager.py

```diff
@@ -0,0 +1,188 @@
+"""Storage manager"""
+from datetime import timedelta
+import logging
+from typing import List, Optional
+
+from django.conf import settings
+from django.db import transaction
+from django.utils.timezone import now
+
+from resolwe.storage.models import FileStorage, StorageLocation, AccessLog
+from resolwe.storage.connectors import connectors, Transfer
+
+
+logger = logging.getLogger(__name__)
+
+
+class DecisionMaker:
+    """Decide which data should be copied to another location or deleted."""
+
+    def __init__(self, file_storage: FileStorage):
+        self.file_storage = file_storage
+
+    def _get_rules(self, connector_name: str, _type: str) -> dict:
+        """Get rules for the given connector name and _type.
+
+        Known rule types are 'copy' and 'delete'.
+        """
+        location_settings = settings.STORAGE_CONNECTORS.get(connector_name, {})
+        return location_settings.get("config", {}).get(_type, {})
+
+    def copy(self) -> List[str]:
+        """Get a list of connector names where data must be copied to."""
+        return [
+            connector_name
+            for connector_name in settings.STORAGE_CONNECTORS
+            if self._should_copy_to(connector_name)
+        ]
+
+    def _should_copy_to(self, connector_name: str) -> bool:
+        """Should data be copied to the location managed by connector name.
+
+        :param connector_name: name of the storage connector.
+        :type connector_name: str
+
+        :return: True when data should be copied to the location managed by
+            connector connector_name.
+        """
+        copy_rules = self._get_rules(connector_name, "copy")
+
+        rule_results = {}
+
+        rule_results["has_rules"] = len(copy_rules) > 0
+
+        rule_results["copy_not_present"] = not self.file_storage.has_storage_location(
+            connector_name
+        )
+
+        if "delay" in copy_rules:
+            move_delay = timedelta(days=copy_rules["delay"])
+            current_delay = now() - self.file_storage.created
+            rule_results["delay"] = current_delay >= move_delay
+
+        return all(rule_results.values())
+
+    def delete(self) -> Optional[StorageLocation]:
+        """Return StorageLocation object that can be deleted.
+
+        When multiple StorageLocation objects can be deleted the one with
+        lowest priority is returned. When no StorageLocation objects can be
+        deleted None is returned.
+
+        :return: storage location that can be safely removed or None.
+        :rtype: Optional[StorageLocation]
+        """
+        can_delete = [
+            (connectors[storage_location.connector_name].priority, storage_location)
+            for storage_location in self.file_storage.storage_locations.all()
+            if self._should_delete(storage_location) and not storage_location.locked
+        ]
+        if can_delete:
+            return sorted(can_delete, key=lambda e: e[0])[-1][1]
```

Comment:
I don't like double indexes :) `connectors[storage_location.connector_name].priority` is only used for sorting, so you can move it into the lambda (or make a separate function for the `key`).

Reply:
OK.